### ISSUE TITLE
Suggestion: Detect `derive(Deserialize)` on type that has methods using `unsafe`

### ISSUE THREAD

**Issue by thomcc** [rust-lang/rust-clippy#5471]
Potential for false positives, but in the past I saw something like:

```rust
#[derive(Clone, ..., Deserialize)]
pub struct NamespacedName {
    // `foo/bar` stored together
    inner: Box<str>,
    // the position of the slash -- constructor ensures it's in bounds
    split: usize,
}

impl NamespacedName {
    #[inline]
	pub fn parts(&self) -> (&str, &str) {
        unsafe { 
            let ns = self.inner.get_unchecked(..self.split);
            let name = self.inner.get_unchecked((self.split + 1)..);
            (ns, name)
        }
    }
}
```

Of course, while the constructor did ensure it was in bounds, `serde` added a second constructor which did not. I suspect other code could have a similar problem.

**Comment by ebroto** [comment]
I would like to work on this one
